Fix autocompetion e2e test

```
FAILED neuro-cli/tests/e2e/test_e2e_blob_storage.py::test_blob_autocomplete - AssertionError: assert 'folder/' in ['some_dir/']
```
https://github.com/neuro-inc/neuro-cli/runs/6325186214?check_suite_focus=true

--- a/neuro-cli/src/neuro_cli/click_types.py
+++ b/neuro-cli/src/neuro_cli/click_types.py
@@ -770,6 +770,7 @@ class PathURLCompleter(URLCompleter, abc.ABC):
                         continue
                     if not item.is_dir() and not self._complete_file:
                         continue
+
                     yield self._make_item(
                         dir_uri,
                         item.name,
@@ -810,20 +811,16 @@ class StoragePathURLCompleter(PathURLCompleter):
 
 class BlobPathURLCompleter(PathURLCompleter):
     async def _is_valid_dir(self, root: Root, uri: URL) -> bool:
-        try:
-            return await root.client.buckets.blob_is_dir(uri)
-        except ResourceNotFound:
-            return False
+        # Not used
+        raise NotImplementedError
 
     @asyncgeneratorcontextmanager
     async def _iter_dir(
         self, root: Root, uri: URL
     ) -> AsyncIterator[PathURLCompleter.DirEntry]:
-        async with root.client.buckets.list_blobs(uri) as it:
-            async for blob_entry in it:
-                if blob_entry.uri == uri:
-                    continue  # Directory itself is also listed as it is prefix search
-                yield blob_entry
+        # Not used
+        raise NotImplementedError
+        yield
 
     async def get_completions(
         self,
@@ -845,14 +842,31 @@ class BlobPathURLCompleter(PathURLCompleter):
                     names = [bucket.id] + ([bucket.name] if bucket.name else [])
                     for name in names:
                         if str(prefix / name).startswith(incomplete):
-                            yield self._make_item(
-                                prefix,
-                                name,
-                                True,
-                            )
+                            yield self._make_item(prefix, name, True)
         else:
-            async for item in super().get_completions(uri, root, incomplete):
-                yield item
+            # Generic get_completions() is not used here because we can
+            # benefit from prefix search in list_blobs().
+            if incomplete.endswith("/"):
+                prefix = uri
+                full_uri = full_uri / ""
+                skip_uri_len = len(full_uri.parts)
+            else:
+                prefix = uri.parent
+                skip_uri_len = None
+
+            async with root.client.buckets.list_blobs(full_uri) as it:
+                async for item in it:
+                    if item.is_dir():
+                        if not self._complete_dir:
+                            continue
+                        # Directory itself is also listed as it is prefix search
+                        if len(item.uri.parts) == skip_uri_len:
+                            continue
+                    else:
+                        if not self._complete_file:
+                            continue
+
+                    yield self._make_item(prefix, item.name, item.is_dir())
 
 
 class PlatformURIType(AsyncType[URL]):

--- a/neuro-cli/tests/unit/test_shell_completion.py
+++ b/neuro-cli/tests/unit/test_shell_completion.py
@@ -322,8 +322,8 @@ def test_blob_autocomplete(run_autocomplete: _RunAC) -> None:
                 id="bucket-1",
                 name="neuro-my-bucket",
                 created_at=datetime(2018, 1, 1, 3),
-                cluster_name="test-cluster",
-                owner="test-user",
+                cluster_name="default",
+                owner="user",
                 provider=Bucket.Provider.AWS,
                 imported=False,
                 org_name=None,
@@ -332,7 +332,7 @@ def test_blob_autocomplete(run_autocomplete: _RunAC) -> None:
                 id="bucket-2",
                 name="neuro-public-bucket",
                 created_at=datetime(2018, 1, 1, 17, 2, 4),
-                cluster_name="test-cluster",
+                cluster_name="default",
                 owner="public",
                 provider=Bucket.Provider.AWS,
                 imported=False,
@@ -342,7 +342,7 @@ def test_blob_autocomplete(run_autocomplete: _RunAC) -> None:
                 id="bucket-3",
                 name="neuro-shared-bucket",
                 created_at=datetime(2018, 1, 1, 13, 1, 5),
-                cluster_name="test-cluster",
+                cluster_name="default",
                 owner="another-user",
                 provider=Bucket.Provider.AWS,
                 imported=False,
@@ -356,33 +356,55 @@ def test_blob_autocomplete(run_autocomplete: _RunAC) -> None:
         async def list_blobs(uri: URL) -> AsyncIterator[BlobObject]:
             async with list(uri.host) as it:
                 async for bucket in it:
+                    try:
+                        key = bucket.get_key_for_uri(uri)
+                    except ValueError:
+                        continue
                     break
                 else:
                     return
-            yield BlobObject(
-                key="file1024.txt",
-                modified_at=datetime(2018, 1, 1, 14, 0, 0),
-                bucket=bucket,
-                size=1024,
-            )
-            yield BlobObject(
-                key="file_bigger.txt",
-                modified_at=datetime(2018, 1, 1, 14, 0, 0),
-                bucket=bucket,
-                size=1_024_001,
-            )
-            yield BlobObject(
-                key="folder2/info.txt",
-                modified_at=datetime(2018, 1, 1, 14, 0, 0),
-                bucket=bucket,
-                size=240,
-            )
-            yield BlobObject(
-                key="folder2/",
-                modified_at=datetime(2018, 1, 1, 14, 0, 0),
-                bucket=bucket,
-                size=0,
-            )
+            blobs = [
+                BlobObject(
+                    key="file1024.txt",
+                    modified_at=datetime(2018, 1, 1, 14, 0, 0),
+                    bucket=bucket,
+                    size=1024,
+                ),
+                BlobObject(
+                    key="otherfile.txt",
+                    modified_at=datetime(2018, 1, 1, 14, 0, 0),
+                    bucket=bucket,
+                    size=1024,
+                ),
+                BlobObject(
+                    key="file_bigger.txt",
+                    modified_at=datetime(2018, 1, 1, 14, 0, 0),
+                    bucket=bucket,
+                    size=1_024_001,
+                ),
+                BlobObject(
+                    key="folder2/info.txt",
+                    modified_at=datetime(2018, 1, 1, 14, 0, 0),
+                    bucket=bucket,
+                    size=240,
+                ),
+                BlobObject(
+                    key="folder2/",
+                    modified_at=datetime(2018, 1, 1, 14, 0, 0),
+                    bucket=bucket,
+                    size=0,
+                ),
+                BlobObject(
+                    key="folder23/",
+                    modified_at=datetime(2018, 1, 1, 14, 0, 0),
+                    bucket=bucket,
+                    size=0,
+                ),
+            ]
+            for blob in blobs:
+                if blob.key.startswith(key):
+                    if "/" not in blob.key[len(key) :].rstrip("/"):
+                        yield blob
 
         mocked_list.side_effect = list
         mocked_blob_is_dir.side_effect = blob_is_dir
@@ -436,27 +458,31 @@ def test_blob_autocomplete(run_autocomplete: _RunAC) -> None:
         zsh_out, bash_out = run_autocomplete(["blob", "ls", "blob:bucket-1/"])
         assert bash_out == (
             "uri,file1024.txt,bucket-1/\n"
+            "uri,otherfile.txt,bucket-1/\n"
             "uri,file_bigger.txt,bucket-1/\n"
-            "uri,info.txt,bucket-1/\n"
-            "uri,folder2/,bucket-1/"
+            "uri,folder2/,bucket-1/\n"
+            "uri,folder23/,bucket-1/"
         )
         assert zsh_out == (
             "uri\nfile1024.txt\n_\nblob:bucket-1/\n"
+            "uri\notherfile.txt\n_\nblob:bucket-1/\n"
             "uri\nfile_bigger.txt\n_\nblob:bucket-1/\n"
-            "uri\ninfo.txt\n_\nblob:bucket-1/\n"
-            "uri\nfolder2/\n_\nblob:bucket-1/"
+            "uri\nfolder2/\n_\nblob:bucket-1/\n"
+            "uri\nfolder23/\n_\nblob:bucket-1/"
         )
 
         zsh_out, bash_out = run_autocomplete(["blob", "ls", "blob:bucket-1/f"])
         assert bash_out == (
             "uri,file1024.txt,bucket-1/\n"
             "uri,file_bigger.txt,bucket-1/\n"
-            "uri,folder2/,bucket-1/"
+            "uri,folder2/,bucket-1/\n"
+            "uri,folder23/,bucket-1/"
         )
         assert zsh_out == (
             "uri\nfile1024.txt\n_\nblob:bucket-1/\n"
             "uri\nfile_bigger.txt\n_\nblob:bucket-1/\n"
-            "uri\nfolder2/\n_\nblob:bucket-1/"
+            "uri\nfolder2/\n_\nblob:bucket-1/\n"
+            "uri\nfolder23/\n_\nblob:bucket-1/"
         )
 
         zsh_out, bash_out = run_autocomplete(["blob", "ls", "blob:bucket-1/fi"])
@@ -469,22 +495,14 @@ def test_blob_autocomplete(run_autocomplete: _RunAC) -> None:
         )
 
         zsh_out, bash_out = run_autocomplete(["blob", "ls", "blob:bucket-1/folder2"])
-        assert bash_out == "uri,folder2/,bucket-1/"
-        assert zsh_out == "uri\nfolder2/\n_\nblob:bucket-1/"
+        assert bash_out == ("uri,folder2/,bucket-1/\n" "uri,folder23/,bucket-1/")
+        assert zsh_out == (
+            "uri\nfolder2/\n_\nblob:bucket-1/\n" "uri\nfolder23/\n_\nblob:bucket-1/"
+        )
 
         zsh_out, bash_out = run_autocomplete(["blob", "ls", "blob:bucket-1/folder2/"])
-        assert bash_out == (
-            "uri,file1024.txt,bucket-1/folder2/\n"
-            "uri,file_bigger.txt,bucket-1/folder2/\n"
-            "uri,info.txt,bucket-1/folder2/\n"
-            "uri,folder2/,bucket-1/folder2/"
-        )
-        assert zsh_out == (
-            "uri\nfile1024.txt\n_\nblob:bucket-1/folder2/\n"
-            "uri\nfile_bigger.txt\n_\nblob:bucket-1/folder2/\n"
-            "uri\ninfo.txt\n_\nblob:bucket-1/folder2/\n"
-            "uri\nfolder2/\n_\nblob:bucket-1/folder2/"
-        )
+        assert bash_out == "uri,info.txt,bucket-1/folder2/"
+        assert zsh_out == "uri\ninfo.txt\n_\nblob:bucket-1/folder2/"
 
 
 def make_job(
@@ -541,8 +559,6 @@ def test_job_autocomplete(run_autocomplete: _RunAC) -> None:
             cluster_name: Optional[str] = None,
             owners: Iterable[str] = (),
         ) -> AsyncIterator[JobDescription]:
-            # print(f"cluster_name = {cluster_name}")
-            # print(f"owners = {owners}")
             for job in jobs:
                 if cluster_name and job.cluster_name != cluster_name:
                     continue

